pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -450,17 +494,60 @@ func (configgen *ConfigGeneratorImpl) buildSidecarInboundListenerForPortOrUDS(no
 		// add one empty entry to the list so we generate a default listener below
 		allChains = []plugin.FilterChain{{}}
 	}
-	for _, chain := range allChains {
+
+	// Detect protocol by sniffing and double the filter chain
+	if pluginParams.ListenerProtocol == plugin.ListenerProtocolAuto {
+		allChains = append(allChains, allChains...)
+		listenerOpts.needHTTPInspector = true
+	}
+
+	for id, chain := range allChains {
 		var httpOpts *httpListenerOpts
 		var tcpNetworkFilters []*listener.Filter
+		var filterChainMatch *listener.FilterChainMatch
 
 		switch pluginParams.ListenerProtocol {
 		case plugin.ListenerProtocolHTTP:
+			filterChainMatch = chain.FilterChainMatch
 			httpOpts = configgen.buildSidecarInboundHTTPListenerOptsForPortOrUDS(node, pluginParams)
 
 		case plugin.ListenerProtocolTCP:
+			filterChainMatch = chain.FilterChainMatch
 			tcpNetworkFilters = buildInboundNetworkFilters(pluginParams.Env, pluginParams.Node, pluginParams.ServiceInstance)
 
+		case plugin.ListenerProtocolAuto:
+			// TODO(crazyxy) avoid bypassing authN using TCP
```

Comment:
this is a pretty big todo, can we make sure this is documented? We need to make it clear this is not prod ready/secure yet

Reply:
I will open an issue for this.